protoquant adding linear primitives

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #24
* __->__ #22

Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757987](https://our.internmc.facebook.com/intern/diff/D45757987)

HDCharles committed:
protoquant adding linear primitives

Summary: WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on ao_experimental/quant_primitives.py:
```diff
@@ -178,6 +178,222 @@ def dequantize_per_channel(
     return y
 
 
+class DynamicallyQuantizedLinear(torch.nn.Module):
```
nit: I'd recommend a separate file for modules, this is a user facing API which is separate from primitives

HDCharles replied:
ok, will do

vkuzo commented on ao_experimental/quant_primitives.py:
```diff
@@ -178,6 +178,222 @@ def dequantize_per_channel(
+    #
+
+    tmp = x_vals_int8.reshape(-1, x_vals_int8.shape[-1])
+    XW_int32 = safe_int_mm(tmp, w_int8_t)
```
nice!

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757987).

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757987](https://our.internmc.facebook.com/intern/diff/D45757987)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757987).

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757987](https://our.internmc.facebook.com/intern/diff/D45757987)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757987).

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757987](https://our.internmc.facebook.com/intern/diff/D45757987)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757987).

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757987](https://our.internmc.facebook.com/intern/diff/D45757987)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757987).

vkuzo commented on ao_experimental/quantized_modules.py:
```diff
@@ -0,0 +1,100 @@
+from quant_primitives import dynamically_quantize_per_channel, quant_int8_dynamic_linear
+
+
+class DynamicallyQuantizedLinear(torch.nn.Module):
```
I'd probably make sure that it's clear this is not a final API and it may change at any point without warning - let's add to docblock?  We haven't really designed the user facing side of this yet.

HDCharles replied:
sounds good

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757987](https://our.internmc.facebook.com/intern/diff/D45757987)

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757987](https://our.internmc.facebook.com/intern/diff/D45757987)

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757987](https://our.internmc.facebook.com/intern/diff/D45757987)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757987).

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757987](https://our.internmc.facebook.com/intern/diff/D45757987)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757987).

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757987](https://our.internmc.facebook.com/intern/diff/D45757987)

[ghstack-poisoned]

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757987](https://our.internmc.facebook.com/intern/diff/D45757987)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757987).

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757987](https://our.internmc.facebook.com/intern/diff/D45757987)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757987).

HDCharles committed:
Update on "protoquant adding linear primitives"


Summary: adding linear primitives, these are adapted from vkuzo's
ao_benchmarks repo. The main change is that usage of safe_int_mm allows
for less complex code because we don't have to keep track of whether
weight is transposed or not. We also implement
DynamicallyQuantizedLinear to ease usage of these primitives and allow
from_float to easily swap from fp32 linear to DynamicallyQuantizedLinear
which calls both the equivalent functional quant_int8_dynamic_linear and
quant_int8_matmul

Test Plan: python test_quant_primitives.py TestEndToEnd
testQuantInt8MatMul TestQuantInt8DynamicLinearOp

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757987](https://our.internmc.facebook.com/intern/diff/D45757987)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757987).
